libbpf-rs/src/user_ringbuf.rs

```diff
@@ -0,0 +1,171 @@
+    /// to submitting the sample.
+    pub fn as_mut(&mut self) -> &mut T {
+        unsafe { &mut *(self.ptr.as_ptr()) }
+    }
+}
+
+impl<T> Drop for UserRingBufferSample<'_, T> {
+    fn drop(&mut self) {
+        // If the sample has not been submitted, explicitly discard it.
+        // This is necessary to avoid leaking ring buffer memory.
+        if !self.submitted {
+            unsafe {
+                libbpf_sys::user_ring_buffer__discard(
+                    self.rb.ptr.as_ptr(),
+                    self.ptr.as_ptr() as *mut _,
+                );
+            }
+        }
+    }
+}
+
+/// Represents a user ring buffer. This is a special kind of map that is used to
+/// transfer data between user space and kernel space.
+#[derive(Debug)]
+pub struct UserRingBuffer {
+    // A non-null pointer to the underlying user ring buffer.
+    ptr: NonNull<libbpf_sys::user_ring_buffer>,
+}
+
+impl UserRingBuffer {
+    /// Create a new user ring buffer from a map.
+    ///
+    /// # Errors
+    /// * If the map is not a user ring buffer.
+    /// * If the underlying libbpf function fails.
+    pub fn new(map: &MapHandle) -> Result<Self> {
+        if map.map_type() != MapType::UserRingBuf {
+            return Err(Error::with_invalid_data("Must use a UserRingBuf map"));
+        }
+
+        let fd = map.as_fd();
+        let raw_ptr = unsafe { libbpf_sys::user_ring_buffer__new(fd.as_raw_fd(), null_mut()) };
+
+        let ptr = NonNull::new(raw_ptr).ok_or_else(|| {
+            // Safely get the last OS error after a failed call to user_ring_buffer__new
+            let errno = unsafe { *libc::__errno_location() };
+            Error::from_raw_os_error(errno)
+        })?;
+
+        Ok(UserRingBuffer { ptr })
+    }
+
+    /// Reserve space in the ring buffer for a sample of type `T``.
+    ///
+    /// Returns a [`UserRingBufferSample<T>`](UserRingBufferSample<'self,T>)
+    /// that contains a mutable reference to a type T that can be written to.
+    /// The sample must be submitted via [`UserRingBuffer::submit`] before it is
+    /// dropped.
+    ///
+    /// This function is *not* thread-safe. It is necessary to synchronize
+    /// amongst multiple producers when invoking this function.
+    pub fn reserve<T>(&self) -> Result<UserRingBufferSample<'_, T>>
+    where
+        T: Sized,
+    {
+        let size = size_of::<T>();
+
+        let sample_ptr =
+            unsafe { libbpf_sys::user_ring_buffer__reserve(self.ptr.as_ptr(), size as c_uint) }
+                as *mut T;
+
+        let ptr = NonNull::new(sample_ptr).ok_or_else(|| {
+            // Fetch the current value of errno to determine the type of error.
+            let errno = unsafe { *libc::__errno_location() };
+            match errno {
+                E2BIG => Error::with_invalid_data("Requested size is too large"),
+                ENOSPC => Error::with_invalid_data("Not enough space in the ring buffer"),
```

Comment:
Nit: Can we use lower case error messages, please?
```suggestion
                E2BIG => Error::with_invalid_data("requested size is too large"),
                ENOSPC => Error::with_invalid_data("not enough space in the ring buffer"),
```